feg/gateway/services/n7_n40_proxy/n7/n7.go

```diff
@@ -39,3 +50,509 @@ func NewN7Client(cfg *PCFConfig) (*n7_sbi.ClientWithResponses, error) {
+		return redirProto
+	}
+	if redirInfo.RedirectEnabled != nil {
+		redirProto.Support = protos.RedirectInformation_DISABLED
+		if *redirInfo.RedirectEnabled {
+			redirProto.Support = protos.RedirectInformation_ENABLED
+		}
+	}
+	if redirInfo.RedirectAddressType != nil {
+		redirProto.AddressType = GetRedirectAddrTypeProto(redirInfo.RedirectAddressType)
+	}
+	if redirInfo.RedirectServerAddress != nil {
+		redirProto.ServerAddress = *redirInfo.RedirectServerAddress
+	}
+
+	return redirProto
+}
+
+func GetRedirectAddrTypeProto(
+	redirAddrType *n7_sbi.RedirectAddressType,
+) protos.RedirectInformation_AddressType {
+	if redirAddrType == nil {
+		return protos.RedirectInformation_IPv4
+	}
+	switch *redirAddrType {
+	case n7_sbi.RedirectAddressTypeIPV4ADDR:
+		return protos.RedirectInformation_IPv4
+	case n7_sbi.RedirectAddressTypeIPV6ADDR:
+		return protos.RedirectInformation_IPv6
+	case n7_sbi.RedirectAddressTypeSIPURI:
+		return protos.RedirectInformation_SIP_URI
+	case n7_sbi.RedirectAddressTypeURL:
+		return protos.RedirectInformation_URL
+	default:
+		return protos.RedirectInformation_IPv4
+	}
+}
+
+func GetFlowQosProto(qosData *n7_sbi.QosData) *protos.FlowQos {
+	if qosData == nil {
+		return nil
+	}
+	return &protos.FlowQos{
+		MaxReqBwDl: BitRateStringToUint32(qosData.MaxbrDl),
+		MaxReqBwUl: BitRateStringToUint32(qosData.MaxbrUl),
+		GbrUl:      BitRateStringToUint32(qosData.GbrUl),
+		GbrDl:      BitRateStringToUint32(qosData.GbrDl),
+		Qci:        N5qiToQci(qosData.N5qi),
+	}
+}
+
+func GetFlowQosFromAuthDefQos(defQos *n7_sbi.AuthorizedDefaultQos) *protos.FlowQos {
+	if defQos == nil {
+		return nil
+	}
+	return &protos.FlowQos{
+		MaxReqBwDl: BitRateStringToUint32(defQos.MaxbrDl),
+		MaxReqBwUl: BitRateStringToUint32(defQos.MaxbrUl),
+		GbrUl:      BitRateStringToUint32(defQos.GbrUl),
+		GbrDl:      BitRateStringToUint32(defQos.GbrDl),
+		Qci:        N5qiToQci(defQos.N5qi),
+	}
+}
+
+func BitRateStringToUint32(br *sbi.BitRateRm) uint32 {
+	if br == nil {
+		return 0
+	}
+	convInt, _ := strconv.ParseUint(string(*br), 10, 32)
+	return uint32(convInt)
+}
+
+func N5qiToQci(n5Qi *sbi.N5Qi) protos.FlowQos_Qci {
+	if n5Qi == nil {
+		return protos.FlowQos_QCI_0
+	}
+	return protos.FlowQos_Qci(*n5Qi)
+}
+
+func GetEventTriggersInfoProto(
+	triggers *[]n7_sbi.PolicyControlRequestTrigger,
+	revalTimeout *time.Time,
+) ([]protos.EventTrigger, *timestamp.Timestamp) {
+	var protoRevalidationTime *timestamp.Timestamp
+	protoTriggers := []protos.EventTrigger{}
+	if triggers == nil {
+		return protoTriggers, protoRevalidationTime
+	}
+	for _, trigger := range *triggers {
+		switch trigger {
+		case n7_sbi.PolicyControlRequestTriggerRETIMEOUT:
+			protoTriggers = append(protoTriggers, protos.EventTrigger_REVALIDATION_TIMEOUT)
+			protoRevalidationTime = ConvertToProtoTimeStamp(revalTimeout)
+		}
+	}
+	return protoTriggers, protoRevalidationTime
+}
+
+func GetUsageMonitorsProto(
+	request *protos.CreateSessionRequest,
+	tgppCtx *protos.TgppContext,
```

Comment:
what is this for? I think i remember that was something related with gx/gy but not sure if this will be needed in here. 
Can you investigate in the code to understand what this means please?

Reply:
Usage monitors are set by the PCF/PCRF for reporting Usage when the set volume thresholds are reached. SessionD sends the UsageMonitoringReport in the UpdateSession and also in the TerminateSession.